packages/logging/src/logging-and-telemetry.ts

```diff
@@ -0,0 +1,785 @@
+import type {
+  MongoshBus,
+  ApiEventWithArguments,
+  ApiEvent,
+  UseEvent,
+  EvaluateInputEvent,
+  ShowEvent,
+  ConnectEvent,
+  ScriptLoadFileEvent,
+  StartLoadingCliScriptsEvent,
+  StartMongoshReplEvent,
+  GlobalConfigFileLoadEvent,
+  CryptLibrarySkipEvent,
+  CryptLibraryFoundEvent,
+  SnippetsCommandEvent,
+  SnippetsErrorEvent,
+  SnippetsFetchIndexErrorEvent,
+  SnippetsFetchIndexEvent,
+  SnippetsLoadedEvent,
+  SnippetsLoadSnippetEvent,
+  SnippetsNpmDownloadActiveEvent,
+  SnippetsNpmDownloadFailedEvent,
+  SnippetsNpmLookupEvent,
+  SnippetsRunNpmEvent,
+  SnippetsTransformErrorEvent,
+  EditorRunEditCommandEvent,
+  EditorReadVscodeExtensionsDoneEvent,
+  EditorReadVscodeExtensionsFailedEvent,
+  FetchingUpdateMetadataEvent,
+  FetchingUpdateMetadataCompleteEvent,
+  SessionStartedEvent,
+  MongoshBusEventsMap,
+} from '@mongosh/types';
+import { inspect } from 'util';
+import { MongoLogWriter } from 'mongodb-log-writer';
+import { mongoLogId } from 'mongodb-log-writer';
+import type {
+  AnalyticsIdentifyMessage,
+  MongoshAnalytics,
+  MongoshAnalyticsIdentity,
+} from './analytics-helpers';
+import { hookLogger } from '@mongodb-js/devtools-connect';
+import { MultiSet, toSnakeCase } from './helpers';
+import { Writable } from 'stream';
+
+export class MongoshLoggingAndTelemetry {
+  log: MongoLogWriter;
+  pendingLogEvents: CallableFunction[] = [];
+
+  telemetryAnonymousId: string | undefined;
+  userId: string | undefined;
+  trackingProperties: {
+    mongosh_version: string;
+    session_id: string;
+  };
+  userTraits: AnalyticsIdentifyMessage['traits'] & {
+    [key: string]: unknown;
+  };
+
+  isSetup = false;
+  hasMongoLogWriterInitialized = false;
+
+  constructor(
+    public bus: MongoshBus,
+    public analytics: MongoshAnalytics,
+    public providedTraits: { platform: string } & {
+      [key: string]: unknown;
+    },
+    mongoshVersion: string
+  ) {
+    const dummySink = new Writable({
+      write(chunk, encoding, callback) {
+        callback();
+      },
+    });
+    this.log = new MongoLogWriter('', null, dummySink);
+
+    this.userTraits = {
+      ...providedTraits,
+      session_id: this.log.logId,
+    };
+
+    this.trackingProperties = {
+      mongosh_version: mongoshVersion,
+      session_id: this.log.logId,
+    };
+  }
+
+  public setup() {
+    this._setupBusEventListeners();
+    this.isSetup = true;
+  }
+
+  public setupLogger(logger: MongoLogWriter) {
+    if (!this.isSetup) {
+      throw new Error('Run setup() before setting up the log writer.');
+    }
+    this.log = logger;
+
+    this.userTraits.session_id = this.log.logId;
+    this.trackingProperties.session_id = this.log.logId;
+  }
+
+  public getTelemetryUserIdentity = (): MongoshAnalyticsIdentity => ({
+    anonymousId: this.telemetryAnonymousId ?? (this.userId as string),
+  });
+
+  /** Start processing pending events after mongosh:logger-initialized */
+  private onLoggerInitialized() {
```

Comment:
It seems redundant and fragile that we are attaching a listener on `mongosh:logger-initialized` if we already get informed about the logger being initialized via a direct call to `.setupLogger()` – can we just wrap this method into the latter?

Reply:
to put it in context of the tech design, would `setupLogger` also emit `mongosh:logger-initialized` but we'd not necessarily attach any listeners to it? 

or would we get rid of `mongosh:logger-initialized` altogether?

Reply:
> would `setupLogger` also emit `mongosh:logger-initialized` but we'd not necessarily attach any listeners to it?

This is what I'd lean towards, since "logger initialized" is generally a meaningful event that another part of mongosh could be interested in listening for it. But that's just theoretical, so, no strong feelings about keeping or removing it altogether.